### DISCUSSION TITLE
Is there a reason QuickSpec and it's closures aren't @MainActor by default?

### DISCUSSION THREAD

**Discussion by CraigSiemens** [Quick/Quick#1288] [Q&A] [answered]
I'm working on testing a new type that's annotated `@MainActor` and received a warning in the tests since it's using it in a nonisolated context. That got me wondering why it works for UIKit types that are also annotated with `@MainActor`. It appears like `XCTestCase` methods are always run on the main thread.

I'm wondering if there'd be any issue if `QuickSpec` and the closures used for before/after/it were annotated with `@MainActor` by default?

I've got a branch where I'm trying it out and it appears to be working fine (at least with Quick's tests).
https://github.com/CraigSiemens/Quick/tree/main-actor

**Comment by younata** [comment, marked as answer]
As indicated by my just merging your PR, there's not really any reason not to do this. It wasn't necessary prior to swift concurrency, and no one else had gotten around to fixing it in Quick.

The Async DSL should not be marked as `@MainActor`. But the synchronous DSL can certainly be marked as such.

**Reply by Noobish1** [reply]
This might be a dumb question but why can't the Async DSL be marked as MainActor?

**Reply by CraigSiemens** [reply]
I believe it's to follow the same behaviour as async tests in `XCTestCase` where they'll be run in parallel on many threads. Then your test can opt-in to running on the `MainActor` if needed. 

If they were annotated `@MainActor` by default, there'd be no way to opt-out from them running on the main thread, so it wouldn't be possible to achieve the same level of parallelization as `XCTestCase`.

**Comment by seth-folley** [comment]
I've noted this in the PR ([here](https://github.com/Quick/Quick/pull/1289)) that annotates the sync DSL closures to be `@MainActor`, but noting here as well that annotating all closures used by `QuickSpec` is concerning. Adding the annotations to those closures brings them into the Swift Concurrency domain which seems to break the intent of separating `QuickSpec` and `AsyncSpec`. It feels like if we're adopting Swift Concurrency in our projects (like annotating new types as `@MainActor`), we should be using `AsyncSpec` and have a way to mark those closures as `@MainActor`, where needed, like we can do in `XCTestCase` by annotating test functions as `@MainActor`. 

Since this is closed, I can start a new discussion or issue as needed but wanted to chime in here first with the original participants.

**Reply by seth-folley** [reply]
Issue filed: https://github.com/Quick/Quick/issues/1298
